### PR TITLE
Adds cml- and oml-permissions with directives

### LINKED ISSUE #174: Check new permissions for orga-layer
Check, if all new permissions for the orga-layer is already implemented in the client.

### PR DESCRIPTION
**Pull request by GabrielInTheWorld**
Fixes #174 

Adds also permission-checks to some views. 

There are two remaining questions:
1. Who has access to committees? Every user, who has one of the permissions `[CML.can_manage, OML.can_manage_users, OML.can_manage_organisation, OML.superadmin]`?
2. Who is authorized to "edit" a meeting? Every user, who has one of the permissions `[CML.can_manage (if committee-ids match), OML.can_manage_users, OML.can_manage_organisation, OML.superadmin]`?

### CODE REVIEW

**Comment by FinnStutzenstein** [comment]
> 1. Who has access to committees? Every user, who has one of the permissions `[CML.can_manage, OML.can_manage_users, OML.can_manage_organisation, OML.superadmin]`?

OML.can_manage_organisation and higher, users in `committee/user_ids`.

> 2. Who is authorized to "edit" a meeting? Every user, who has one of the permissions `[CML.can_manage (if committee-ids match), OML.can_manage_users, OML.can_manage_organisation, OML.superadmin]`?

For editing a meeting (same for committee) you can orient yourself from https://github.com/OpenSlides/OpenSlides/wiki/meeting.update The form must present the fields as given in the permission-section

File: client/src/app/core/core-services/operator.service.ts [hunk 1]
```diff
@@ -26,12 +27,12 @@ function getUserCML(user: ViewUser): { [id: number]: string } | null {
         return null;
     }
 
-    const CML = {};
+    const permissionObject = {};
     const templateField = (user as any).committee_$_management_level;
     for (const replacement of templateField) {
-        CML[+replacement] = (user as any)[`committee_$${replacement}_management_level`];
+        permissionObject[+replacement] = (user as any)[`committee_$${replacement}_management_level`];
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:33]: You should be able to use `user.committee_manegement_level(id)`. This was a provisorium of mine due to the fact that the user didn't had a CML back then.

File: client/src/app/core/core-services/operator.service.ts [hunk 2]
```diff
@@ -26,12 +27,12 @@ function getUserCML(user: ViewUser): { [id: number]: string } | null {
         return null;
     }
 
-    const CML = {};
+    const permissionObject = {};
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:30]: The name is too generic (Which permission?). I like the old name better.

File: client/src/app/core/core-services/organization-permission.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+export enum OML {
+    superadmin = 'superadmin',
+    can_manage_organisation = 'can_manage_organisation',
+    can_manage_users = 'can_manage_users',
+    default = 'default'
+}
+
+export const permsLookup = {
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/organization-permission.ts:13]: make two exported consts out of this and name it properly like OmlNameMapping

File: client/src/app/core/core-services/operator.service.ts [hunk 4]
```diff
@@ -415,6 +416,24 @@ export class OperatorService {
         return result;
     }
 
+    public hasOrganizationPermissions(...permissionsToCheck: OML[]): boolean {
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:419]: docs (same below): What does it mean to give multiple: and'ed or or'd?

File: client/src/app/core/core-services/operator.service.ts [hunk 5]
```diff
@@ -415,6 +416,24 @@ export class OperatorService {
...
+        return permissionsToCheck.every(permission => (permsLookup.oml[this.OML] || 0) >= permsLookup.oml[permission]);
+    }
+
+    public hasCommitteePermissions(committeeId: Id, ...permissionsToCheck: CML[]): boolean {
+        // If a user can manage an entire organisation, he can also manage every committee.
+        const orgaPermissions = permissionsToCheck.includes(CML.can_manage) ? [OML.can_manage_organisation] : [];
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:425]: Is there a usecase for this swithc? I would expect to always check for OML.can_manage_organization or higher regardless of the CML requested.

File: client/src/app/management/components/committee-list/committee-list.component.html [hunk 6]
```diff
@@ -95,6 +95,7 @@ <h2>{{ 'Committees' | translate }}</h2>
 
     <div *pblNgridCellDef="'menu'; row as committee" class="cell-slot fill">
         <button
+            *osCmlPerms="CML.can_manage; committeeId: committee.id"
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/committee-list/committee-list.component.html:98]: the syntax of this directive looks good

File: client/src/app/management/components/meeting-list/meeting-list.component.html [hunk 7]
```diff
@@ -2,7 +2,7 @@
     [customMenu]="true"
     prevUrl="/committees"
     [nav]="false"
-    [hasMainButton]="true"
+    [hasMainButton]="operator.hasCommitteePermissions(currentCommittee?.id, CML.can_manage)"
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/meeting-list/meeting-list.component.html:5]: as I see the `?`: Please explicitly add `null|undefined` to the method in the operator and handle null cases explicitly. I know that it currently works, but handling these things explicitly gives clearance in the behavior.

File: client/src/app/shared/directives/base-perms.directive.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
+import { Directive, OnDestroy, OnInit, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import { Subscription } from 'rxjs';
+
+import { OperatorService } from 'app/core/core-services/operator.service';
+
+@Directive()
+export abstract class BasePermsDirective<P> implements OnInit, OnDestroy {
```
Comment (FinnStutzenstein) [inline client/src/app/shared/directives/base-perms.directive.ts:8]: Good abstraction!

**Review by FinnStutzenstein — CHANGES_REQUESTED** [review]

**Comment by FinnStutzenstein** [comment]
This might be for another PR (if so, please create an issue): Is the authguard capable of handing OML/CML? In a meeting we can give permissions for each router entry. This will be needed too für orga things like the global user list

File: client/src/app/core/core-services/operator.service.ts (hunk 5 above)
Comment (GabrielInTheWorld) [reply client/src/app/core/core-services/operator.service.ts:425]: In the docs the permissions are described as follows: 
```
A user with superadmin can do everything: He is allowed to see non-personal models (e.g. not personal notes or password hashes of other users), but can enter every committee and meeting and is transitively manager of each committee and admin of each meeting.

A user with can_manage_organisation is indirectly a committee manager for every committee. The user can create, edit and delete committees and administrating resources and the motion forwarding structure.
```

Therefore, I checked this. But then, I can always check, which OML an operator has.

File: client/src/app/management/components/meeting-list/meeting-list.component.html (hunk 7 above)
Comment (FinnStutzenstein) [reply client/src/app/management/components/meeting-list/meeting-list.component.html:5]: yes, this is right, but if you give `null` or `undefined`, it should return false.

File: client/src/app/core/core-services/operator.service.ts [hunk 9]
```diff
@@ -16,22 +17,23 @@ import { GroupRepositoryService } from '../repositories/users/group-repository.s
...
-    if (!(user as any).committee_$_management_level) {
+    if (!user.committee_$_management_level) {
         return null;
     }
 
-    const CML = {};
-    const templateField = (user as any).committee_$_management_level;
+    const committeeManagementLevel = {};
+    const templateField = user.committee_$_management_level;
     for (const replacement of templateField) {
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:33]:
```suggestion
    for (const replacement of user.committee_$_management_level) {
```

File: client/src/app/core/core-services/operator.service.ts [hunk 10]
```diff
@@ -415,6 +417,49 @@ export class OperatorService {
...
+
+    /**
+     * Determines whether the current operator is included in at least one of the committees, which are passed.
+     *
+     * @param committees Several committees to check if the current operator is included in them.
+     *
+     * @returns `true`, if the current operator is included in at least one of the given committees.
+     */
+    public isInCommittees(...committees: Committee[]): boolean {
+        return committees.some(committee => this.userSubject.value.committee_ids?.includes(committee.id));
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/operator.service.ts:460]: If you are using `committee_ids`, you must make sure, the operator requests this field.

File: client/src/app/management/components/committee-list/committee-list.component.ts [hunk 11]
```diff
@@ -127,4 +141,12 @@ export class CommitteeListComponent extends BaseListViewComponent<ViewCommittee>
             fieldset: []
         };
     }
+
+    private canSeeCommittee(committee: ViewCommittee): boolean {
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/committee-list/committee-list.component.ts:145]: Maybe you mean `canAccess`? This should be done by the autoupdate/restricter service -> You should only get a list of committees you can see. If you want to filter for access (similar to meetings), use the `canAccess` feature in the basemodel.

File: client/src/app/management/components/management-navigation/management-navigation.component.ts [hunk 12]
```diff
@@ -3,50 +3,51 @@ import { MatDialog } from '@angular/material/dialog';
...
     styleUrls: ['./management-navigation.component.scss']
 })
 export class ManagementNavigationComponent {
-    public menuEntries: MainMenuEntry[] = [
+    public menuEntries: OrgaMenuEntry[] = [
         {
             route: '/',
             displayName: 'Dashboard',
             icon: 'dashboard',
-            // permission: Permission.coreCanSeeFrontpage,
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/management-navigation/management-navigation.component.ts:21]: whats about OMLs?

File: client/src/app/management/components/meeting-edit/meeting-edit.component.html [hunk 13]
```diff
@@ -54,6 +54,20 @@ <h2 *ngIf="!isCreateView">{{ editMeetingLabel | translate }}</h2>
...
+            <os-search-value-selector
+                formControlName="userIds"
+                [multiple]="true"
+                placeholder="{{ 'Participants' | translate }}"
+                [inputListValues]="committee?.users"
+            ></os-search-value-selector>
+        </mat-form-field>
+
+        <!-- server bug -->
+        <!-- <mat-checkbox disabled formControlName="set_as_template">Set as template</mat-checkbox> -->
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/meeting-edit/meeting-edit.component.html:67]: please create an issue then.

File: client/src/app/management/components/meeting-edit/meeting-edit.component.html [hunk 14]
```diff
@@ -54,6 +54,20 @@ <h2 *ngIf="!isCreateView">{{ editMeetingLabel | translate }}</h2>
...
+
+        <!-- Permission: CML.can_manage || OML.can_manage_organisation -->
```
Comment (FinnStutzenstein) [inline client/src/app/management/components/meeting-edit/meeting-edit.component.html:69]: does not fit to the permission check below.

**Review by FinnStutzenstein — CHANGES_REQUESTED** [review]

File: client/src/app/management/components/meeting-edit/meeting-edit.component.html (hunk 14 above)
Comment (GabrielInTheWorld) [reply client/src/app/management/components/meeting-edit/meeting-edit.component.html:69]: Why? When checked for any CML-permission, `OML.can_manage_organisation` or higher satisfy also that permission.

File: client/src/app/management/components/meeting-edit/meeting-edit.component.html (hunk 13 above)
Comment (GabrielInTheWorld) [reply client/src/app/management/components/meeting-edit/meeting-edit.component.html:67]: Done in #239

File: client/src/app/management/components/committee-list/committee-list.component.ts (hunk 11 above)
Comment (GabrielInTheWorld) [reply client/src/app/management/components/committee-list/committee-list.component.ts:145]: I've understand this in another way -> but then, you have right.

File: client/src/app/management/components/meeting-edit/meeting-edit.component.html (hunk 14 above)
Comment (FinnStutzenstein) [reply client/src/app/management/components/meeting-edit/meeting-edit.component.html:69]: Yes, you are right. Please adjust the comment to sound like `Permission: CML.can_manage (implicit: or OML.can_manage_organisation or higher)`

**Comment by FinnStutzenstein** [comment]
I've found a modelling issue: You cannot trust only on `user/committee_$_managment_level`. You **must** always additionally query `user/committee_ids` since the following scenario is valid:

A user can have
```
committee_$_managment_level = ["1", "2"]
committee_$1_managment_level = "can_manage"
committee_$2_managment_level = "can_manage"
committee_ids = [1]
```

Querying if the user has manage perms in committee 2 would currently lead to `true`, which is wrong.

This happens due to the fact, that the backend does not update template fields, if the underlying relation changes [1]. It is reproduceable when creating a user, assign it to both committees and setting his CML for each meeting to can_manage. Then remove the user from one meeting.

For the permission check: In the operator check, if he has the CML in committee <id>, you first have to check that the user is assigned to the committee and do not only trust the CML. This requires you to always request `committee_ids` from the autoupdate service. I think I already said that in an earlier comment.

[1] Note, that this is also the case for e.g. `number_$`: There might be a number for a meeting where the user is not assigned to. It is harmless since the client will not show it and it is no security issue. We declared it as a feature: If the user is reassigned to the meeting, he has his old data back ;)

**Review by FinnStutzenstein — APPROVED** [review]
OK for me. Looks good. I could not test it but since these changes are necessary to have them in other PRs and features you can merge it.

**Pull request merged**